### PR TITLE
Proposing solution for: Automate Publishing of Javadocs #2631

### PR DESCRIPTION
**Pull request by rpmcdougall**
This is referencing issue https://github.com/ReactiveX/RxJava/issues/2631 to automate publishing of JavaDocs for 2.x

I've added in a Gradle task using https://github.com/ajoberstar/gradle-git-publish and provided a block to pick up a Github token for auth if present. You can publish a doc with this method by generating the JavaDocs with the already included Gradle task javadocJar and then running the gitPublishPush Gradle task. I have this working here: https://rpmcdougall.github.io/RxJava/

Thanks!

### CODE REVIEW

**Comment by rpmcdougall** [comment]
Looks like I will need to do some work to get it to fit the current javadoc folder structure and properly generate for specific versions in 2.x to match what is currently there.

File: build.gradle [hunk 1]
```diff
@@ -103,6 +107,30 @@ javadoc {
...
+    // branch will be created if it doesn't exist
+    branch = 'gh-pages'
+
+    contents {
+        from 'build/docs/javadoc'
+        from(javadoc) {
+            into 'api'
+        }
+    }
+    commitMessage = 'Publishing a new page'
```
Comment (akarnokd) [inline build.gradle]: Including the version message would be great.

File: build.gradle [hunk 2]
```diff
@@ -103,6 +107,30 @@ javadoc {
     }
 }
 
+
+
+if (project.hasProperty('githubToken')) {
+    System.setProperty('org.ajoberstar.grgit.auth.username', project.githubToken)
+}
+
+gitPublish {
```
Comment (akarnokd) [inline build.gradle]: Can this be conditional on the build mode? We'd like to publish only on release builds.

File: build.gradle [hunk 3]
```diff
@@ -103,6 +107,30 @@ javadoc {
     }
 }
 
+
+
+if (project.hasProperty('githubToken')) {
```
Comment (akarnokd) [inline build.gradle]: How do I create this token value? `-PgithubToken={$javadocGithubToken}` in [buildViaTravis](https://github.com/ReactiveX/RxJava/blob/2.x/gradle/buildViaTravis.sh#L21)?

**Comment by akarnokd** [comment]
Please also make sure it doesn't delete the v1 javadocs by first clearing all the target directories.

**Comment by rpmcdougall** [comment]
I've noted all of your suggestions, and I'll try to get those in order. As far as the token, I was thinking it was more of an optional thing depending on how one would authenticate to the repo. If its just travis that will be doing the publishing, I don't know that its completely necessary.

File: build.gradle [hunk 4]
```diff
@@ -103,6 +107,30 @@ javadoc {
...
+        branch = 'gh-pages-test'
+
+        contents {
+            from 'build/docs'
+        }
+
+        preserve {
+            include '1.x/**'
+            include '2.x/**'
+            include 'images/**'
```
Comment (akarnokd) [inline build.gradle]: There is an `rx` directory here to be preserved as well: https://github.com/ReactiveX/RxJava/tree/gh-pages/javadoc

File: build.gradle [hunk 5]
```diff
@@ -103,6 +107,30 @@ javadoc {
     }
 }
 
+
+if (rootProject.hasProperty("releaseMode")) {
```
Comment (akarnokd) [inline build.gradle]: releaseMode==full

File: build.gradle [hunk 6]
```diff
@@ -103,6 +107,30 @@ javadoc {
...
+    gitPublish {
+
+        repoUri = 'https://github.com/rpmcdougall/RxJava.git'
```
Comment (akarnokd) [inline build.gradle]: I guess this is for testing it in your own repo, right?

Comment (rpmcdougall) [reply build.gradle]: Indeed, it will be changed to the proper repo when everything else is worked out.

**Comment by rpmcdougall** [comment]
So far it appears to be matching up with the current gh-pages branch as far as maintaining what needs to be there. However, I'm not sure the appropriate way to test making sure the release conditional is working as I'm super green when it comes to releases. Hopefully what I've got so far is mostly in line with what you are looking for. 

Here is the latest publish on my repo:
https://github.com/rpmcdougall/RxJava/tree/gh-pages
https://rpmcdougall.github.io/RxJava/javadoc/

I'm also curious how the project is migrating previous release documentation to the 2.x directory since I'm wondering how that might fit into what I am working on here if at all.

**Comment by akarnokd** [comment]
To be clear, in the manual process I copy the files to 3 locations:

- under `/javadoc`, preserving the `rx` and `resources` directory
- under `2.x/javadoc`
- under `2.x/javadoc/2.x.y` release version specific

**Comment by akarnokd** [comment]
I've cobbled together a script in one of my support libraries and it seems to work. I guess custom copying patterns can be made via [these kinds of cps](https://github.com/akarnokd/RxJava2Jdk8Interop/blob/master/push.sh#L41).

**Comment by rpmcdougall** [comment]
I was also just looking into making a couple more Gradle copy specs, but I could also do something liek that as well.
https://docs.gradle.org/current/javadoc/org/gradle/api/file/CopySpec.html

**Comment by rpmcdougall** [comment]
Latest push seems to do the trick except for the root of 2.x/javadoc which doesnt seem to be caught up with 2.19. Are some of those previous releases in that directory not full releases? I believe I was caught between my preserve statements and my copyspec colliding a bit, so I'll have to work on that.

https://github.com/rpmcdougall/RxJava/tree/cpstest/2.x/javadoc

**Comment by akarnokd** [comment]
I started working on a bash script version: https://github.com/ReactiveX/RxJava/tree/JavadocAutoPublish and https://github.com/ReactiveX/RxJava/blob/JavadocAutoPublish/gradle/push.sh . I'm not proficient with shell scripts so it may take some tries to get it working with RxJava as well.

**Comment by akarnokd** [comment]
Thanks for your efforts, but for now, we'll go with the shell script variant, which should give more control over what happens.

**Pull request closed without merging**